### PR TITLE
More fixes to canonicalization

### PR DESCRIPTION
**Pull request by mmccool**
- Discuss use of base.  In particular, add a clarification that the previous assertion about not changing the expression of URLs *includes* not changing how URLs used for interaction endpoints are expressed as a combination of base and relative URLs.  This is also meant to address the point about base raised in https://github.com/w3c/wot-profile/pull/57
- Add an assertion stating that if a prefix is defined, it MUST be used in place of that URL.  This is to support stable round-tripping where we may expand prefixes into URLs and then on export have to "un-expand" them.  If a URL was used in place of a prefix in the input it would be a legal TD, but the unexpansion would replace it with the prefix, which would result in a serialization different than the input.
- <s>Update: No "security" at the Thing level, and added assertion that names of securityDefinitions must be preserved.  This avoids problems with "equivalent" TDs using different names for these definitions.  See comments below.</s> (REVERTED)

Note that some of these choices mean that there will be Canonical TDs defining the same "information" that are considered different.  For instance, "not changing the way URLs are expressed" means that TDs that split up base and (relative) interaction URLs differently will be considered different, even though they actually refer to the same endpoints.  Likewise not changing prefixes means we can have any number of TDs with different prefixes that refer to the same URLs for those vocabularies, and therefore technically have the same semantics.  Therefore the goal is not really to stably canonicalize the semantic information itself, but the *expression* (serialization) of a given set of information.

***
<a href="https://pr-preview.s3.amazonaws.com/mmccool/wot-thing-description/pull/1129.html" title="Last updated on May 28, 2021, 1:27 PM UTC (ec5c17b)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/wot-thing-description/1129/f95085b...mmccool:ec5c17b.html" title="Last updated on May 28, 2021, 1:27 PM UTC (ec5c17b)">Diff</a>

### CODE REVIEW

**Comment by mmccool** [comment]
just realized there there is one other issue with "security" at the Thing level vs. at the Form level.  A TD with the same security (say, "nosec") in all Forms can be expressed in two ways, in fact many more than two.  So to avoid variants, will have to add an assertion that Canonical form expresses all security requirements at the Form level (resolving all overrides first) and does not use "security" at the Thing level.

**Comment by mmccool** [comment]
Don't know if I have to add something about the names of securityDefintions being retained.  In *theory* you could define the same information using different names for securityDefinitions, as long as the mapping was the same (these are flattened before being pulled into an RDF database and some extra work is needed to unflatten them when reserializing, too).   Probably to be safe I should add another assertion that forbids processors from changing these names.

**Comment by mmccool** [comment]
Note that we can make the names of defined objects part of the information model to avoid the disconnect between information and its expression with this last point (and for prefixes).  See the round-tripping discussion linked above.

**Comment by mmccool** [comment]
@sebastiankb although you changed the title to WIP, I feel this PR is ready to merge myself (pending review).  Of course I MAY have overlooked other little corner cases for canonicalization (the nature of the problem...) but don't have any more right now.

**Comment by mmccool** [comment]
Note there is a similar issue with additionalExpectedResponses and schemaDefinitions wrt to canonicalization and not changing the names of schema definitions.  However, I can't add such an assertion here until PR https://github.com/w3c/wot-thing-description/pull/1130 is merged, and can't add it to https://github.com/w3c/wot-thing-description/pull/1130 until this PR is merged.  So we should merge both and then I can fix this gap.  I've created an issue to follow up on this: https://github.com/w3c/wot-thing-description/issues/1150

File: index.html
Comment (sebastiankb) [inline index.html:4422]: we need to grantee that normalized prefixes are used

**Review by sebastiankb — CHANGES_REQUESTED** [review]

File: index.html
Comment (danielpeintner) [inline index.html:4333]: This addition looks contradictory to me. Why should sorting and/or suppressing white-spaces make it more verbose and redundant?

File: index.html
Comment (mmccool) [reply index.html:4374]: OK, let me look into that.   You're right, we changed this; it is mandatory at the top level, and optional at the form level.  So the only thing we have to worry about is people "overriding" it at the Form level with the same scheme mentioned at the top level (which would be semantically equivalent, but redundant...)

File: index.html
Comment (mmccool) [reply index.html:4422]: yes, maintaining prefixes is a TD processor requirements (this includes directories).  This can be done though by treating the @context as a regular object...

**Comment by sebastiankb** [comment]
from today's TD call:
- PR not ready to merge
- @mmccool checks the comments above
- step 5 have to be removed. The solutions should be the same as for the securityDefintions.

File: index.html
Comment (mmccool) [reply index.html:4333]: I will try rewording that.  Suppression of whitespace does not make it more verbose, but other choices may.   Or I may just delete that addition.

File: index.html
Comment (mmccool) [reply index.html:4374]: There are two possible solutions here:
1. Leave the definitions alone.  If "security" appears that the Form level, it must appear in the output after round-tripping, even if it is the same as the object at the "Thing" level.  We would have to forbid TD processors from "optimizing" TDs by removing such redundant definitions.
2. Require removal of redundant definitions.
Based on the discussion, it sounds like people would prefer 1.

Comment (mmccool) [reply index.html]: I applied solution 1, adding an assertion that TD processors must not remove (or add) "security" fields in Forms, even if they are redundant.

File: index.html
Comment (mmccool) [reply index.html]: decided to just take those extra words out.

File: index.html
Comment (mmccool) [reply index.html:4414]: I *think* this was already taken care of in the draft, please confirm and resolve the conversation if so.

**Review by sebastiankb — APPROVED** [review]

**Comment by sebastiankb** [comment]
from today's TD call:
- @mmccool addressed the comments above in latest commit
- decide to merge to have a new baseline about canonicaliziation

**Pull request merged**
